scvi/external/cell2location/_base.py

```diff
@@ -0,0 +1,1060 @@
+                    ),
+                    amortised_plate_sites=amortised_vars,
+                    n_in=n_in,
+                    n_hidden=n_hidden,
+                    data_transform=data_transform,
+                    encoder_kwargs=encoder_kwargs,
+                    single_encoder=single_encoder,
+                )
+            )
+
+        self._get_fn_args_from_batch = self._model._get_fn_args_from_batch
+
+    @property
+    def model(self):
+        return self._model
+
+    @property
+    def guide(self):
+        return self._guide
+
+    @property
+    def is_amortised(self):
+        return self._amortised
+
+    def data_transform_clusters(self):
+        def _data_transform(x):
+            return torch.log1p(torch.cat([x, x @ self.gene_clusters], dim=1))
+
+        return _data_transform
+
+    def data_transform_scale(self):
+        def _data_transform(x):
+            # return (x - self.var_mean) / self.var_std
+            return x / self.var_std
+
+        return _data_transform
+
+
+class TrainSampleMixin:
+    """
+    Reimplementation of cell2location [Kleshchevnikov20]_ model. This mixin class provides methods for:
+
+    - training models using minibatches (standard scVI interface) and using full data (copies data to GPU only once).
+    - computing median and quantiles of the posterior distribution using both direct and amortised inference
+    - generating samples from posterior distribution
+
+    """
+
+    @property
+    def _plan_class(self):
+        return PyroTrainingPlan
+
+    def _train_full_data(
+        self,
+        max_epochs: Optional[int] = None,
+        use_gpu: bool = False,
+        plan_kwargs: Optional[dict] = None,
+        lr: float = 0.01,
+        autoencoding_lr: Optional[float] = None,
+        clip_norm: float = 200,
+        continue_training: bool = True,
+    ):
+        """
+        Private method for training using full data.
+
+        Parameters
+        ----------
+        max_epochs
+            Number of training epochs / iterations
+        use_gpu
+            Bool, use gpu?
+        plan_kwargs
+            Training plan arguments such as optim and loss_fn
+        continue_training
+            When the model is already trained, should calling .train() continue training? (False = restart training)
+
+        Returns
+        -------
+        ELBO history in self.module.history_
+
+        """
+
+        args, kwargs = self.module.model._get_fn_args_full_data(self.adata)
+        gpus, device = parse_use_gpu_arg(use_gpu)
+
+        args = [a.to(device) for a in args]
+        kwargs = {k: v.to(device) for k, v in kwargs.items()}
+        self.to_device(device)
+
+        if not continue_training or not self.is_trained_:
+            # models share param store, make sure it is cleared before training
+            pyro.clear_param_store()
+        # initialise guide params (warmup)
+        self.module.guide(*args, **kwargs)
```

Comment:
the `PyroJitGuideWarmup` can be used here too

Reply:
This is also just a single line, and it will be not necessary when this is addressed https://github.com/YosefLab/scvi-tools/issues/1058